### PR TITLE
Added skip profiler functionality

### LINKED ISSUE #4568: Task List - Mark Store details step as incomplete when skipping the profiler
**User story:**
As a new merchant
I want to be able to easily access the store profiler
So I can complete the store information after skipping it

**Description:** with the addition of a skip link in the profiler (#4566), the Store details step should be marked as incomplete if the user skips the store profiler flow.

I'm not sure if this logic was already implemented on #4096  

**Acceptance Criteria:**
- [ ] Display the Store details step as incomplete if the user skipped the store profiler flow.

**Event tracking:**
Currently, we don't have an event that tracks clicks in the task list, so probably is the easiest way to track if users are clicking on the Store Details step:

Event name: `wcadmin_tasklist_click`
Event prop: `task_name` (value: task name)
Event description: When the user clicks to access a specific task

### PR DESCRIPTION
**Pull request by octaedro**
Fixes #4568-partially

This PR adds the functionality to skip the onboarding profiler. Also, it leaves the `Store details` as incomplete in the `Home` screen.

### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots
skip_profiler1
skip_profiler2

### Detailed test instructions:

To test this functionality I created the branch `add/4568_test` where I added a link to be able to skip the onboarding profiler.
1. Delete the profiler information with these two SQL sentences:
````
UPDATE `wp_options` SET `option_value` = 'a:0:{}' WHERE `wp_options`.`option_name` = 'woocommerce_onboarding_profile';
UPDATE `wp_options` SET `option_value` = 'no' WHERE `wp_options`.`option_name` = 'woocommerce_task_list_complete';
````
2. Activate the profiler setup wizard (going to `/wp-admin/admin.php?page=wc-settings` -> Setup wizard -> Profile setup wizard -> `Enable`).
3. Fill out a couple of onboarding steps.
4. Press `Skip`.
skip_profiler3
5. Verify the task list on the `Home` screen has the `Store details` marked as incomplete.
6. Press `Store details`. The profiler should open again.
7. After pressing `Skip`, the profile setup wizard (step 2) should appear as disabled. Verify it's disabled when the profiler was skipped and enabled when it wasn't.

/cc @samueljseay This PR is related to the issue [4566](https://github.com/woocommerce/woocommerce-admin/issues/4566). I know Sam is working on it.

### Changelog Note:
Dev: Added skip profiler functionality

### CODE REVIEW

File: src/Features/Onboarding.php
Comment (joshuatf) [inline src/Features/Onboarding.php:926]: It looks like we use this logic several times throughout the code (`! profilerCompleted && ! profilerSkipped`).  It would be great if we could somehow consolidate this logic into a single place so it's less error prone if we need to change it again.

I've also been thinking about how this works with being able to access the profiler after completion/skip and I see a couple potential ways we can improve the existing setup here:

1. Introduce a setting `isProfilerEnabled` from the server that contains this logic and allows us to toggle on/off if we need to override the logic to access the profiler.

2. Always allow the profiler to be accessed through its own route (or if some query param like `step` is defined) and let the UI redirect to the profiler if not complete or skipped.

I'm leaning towards option 2 because this gives us a lot of flexibility and would simplify issues like hiding the UI during connection (exposed in https://github.com/woocommerce/woocommerce-admin/pull/4717), but would love more feedback on this idea.

**Review by joshuatf — COMMENTED** [review]
Nice job on this one, Fernando!  I left some thoughts about a possible refactor here related to existing code debt we have.

While these aren't caused by this PR, I think it may be a good opportunity for us to do some tidying while we're here.  Feel free to ping me if I can clarify or help with anything here!

File: src/Features/Onboarding.php
Comment (jeffstieler) [reply src/Features/Onboarding.php:926]: At first glance option 2 makes a lot of sense and seems to provide opportunity to clean up a lot of conditional logic we have - I'm thinking about the "home" route render and the forced postback currently required to revisit the OBW from the home screen. Perhaps the only gotchas might be changing existing Tracks events?

Comment (joshuatf) [reply src/Features/Onboarding.php:926]: Thanks for your insights on this, Jeff!

>I'm thinking about the "home" route render and the forced postback currently required to revisit the OBW from the home screen.

This, the soon-to-be task to revisit the profiler, and oddities behind the connection flow make it seem like the right decision to me as well.

>Perhaps the only gotchas might be changing existing Tracks events?

Good question.  I think the main one that might be `storeprofiler_toggled` (`wcadmin_storeprofiler_toggled`) which would no longer be necessary.  Any others you see @jeffstieler?

@pmcpinto are we currently using this or is there any information you think we might miss out on from this event?  I imagine we could still determine if a user revisits after completion with the other existing events.

Comment (octaedro) [reply src/Features/Onboarding.php]: Thank you @joshuatf and @jeffstieler for your suggestions, I made the changes you mentioned in the commit 74f344f.
As soon as Pedro confirms the things related to the event recording I will add the necessary modifications.

@joshuatf could you give another look at this PR?

**Comment by pmcpinto** [comment]
> @pmcpinto are we currently using this or is there any information you think we might miss out on from this event? I imagine we could still determine if a user revisits after completion with the other existing events.

How is this event going to be affected by this change? If the user completes/dismisses the task list and then goes to Help - Setup Wizard - Profile Setup Wizard - Enable, the event will be triggered, correct?

Just a reminder that we aim to include a Dismiss button in the Store Profiler task (#4733), I'm not sure if it will affect some of the logic that's being implemented in this PR. Tks!

**Comment by octaedro** [comment]
> How is this event going to be affected by this change? If the user completes/dismisses the task list and then goes to Help - Setup Wizard - Profile Setup Wizard - Enable, the event will be triggered, correct?

Hi @pmcpinto,
The answer to your question is `yes`, but also this PR  introduces a couple of changes to the way the OBW is accessed. Before it was only possible to access it after executing the method `reset_profiler` (following the steps you mentioned in your comment). This method records the event `storeprofiler_toggled`. Now it's also possible to access it directly using URL: `wp-admin/admin.php?page=wc-admin&path=%2Fprofiler`.
This means that the OBW could be visited without recording the event `storeprofiler_toggled`.
Do you think we should add or modify any of the events? Maybe the event `storeprofiler_toggled` is not necessary anymore. What do you think?

**Comment by pmcpinto** [comment]
@octaedro thanks for the clarification. I think that we should keep `wcadmin_storeprofiler_toggled`, to understand if the number of users enabling the profiler through the help tab is going to decrease.

And in #4568 I asked to implement additional event tracking around clicking in the task list, which allows checking how many users are accessing the profiler through the task list.

**Comment by octaedro** [comment]
Thank you @pmcpinto 🙂
I will leave `wcadmin_storeprofiler_toggled `.

I think that now this PR is ready for another look before merging. I will put this in Slack to see if anyone has time.

File: client/dashboard/index.js
Comment (psealock) [inline client/dashboard/index.js:41]: I think using React Router to change route can avoid loading the page twice. Here is an example:

https://github.com/woocommerce/woocommerce-admin/blob/e09675584236dceaa1f91fcb64f11b69ad52f9cc/client/layout/controller.js#L294

File: client/homescreen/index.js
Comment (psealock) [inline client/homescreen/index.js:29]: same here about React Router.

File: client/profile-wizard/index.js
Comment (psealock) [inline client/profile-wizard/index.js:231]: Same here.

Also, does it make sense to `await` the return of the profile items request here? Can we just navigate to the homescreen right away?

**Review by psealock — APPROVED** [review]
Nice @octaedro, this is looking great and testing well. I'm just noting a few places where it makes sense to probably use React Router's `getHistory` to push a new url in case changing `window.location` causes a full page refresh. I haven't looked into it and that may not even be cause a full page refresh, so I'm approving.

Then there is a comment about removing an `await` that may not be needed.

File: client/dashboard/index.js
Comment (octaedro) [reply client/dashboard/index.js:46]: Hi @psealock, great idea!
I changed it in the commit ebe323c

File: client/homescreen/index.js
Comment (octaedro) [reply client/homescreen/index.js:34]: I did the same in the commit ebe323c

File: client/profile-wizard/index.js
Comment (octaedro) [reply client/profile-wizard/index.js]: And I changed it here too 😀

Even though it looks like the `await` is useless, it is necessary to be sure the `updateProfileItems` has finished before the redirection. Otherwise, when the home screen is loading, the check could fail and it would redirect the user back to the OBW.

File: client/profile-wizard/index.js
Comment (mattsherman) [inline client/profile-wizard/index.js:226]: If `updateProfileItems` errors, how is it handled?

**Comment by psealock** [comment]
Thanks @octaedro changes look good. @mattsherman [brings up a good point](https://github.com/woocommerce/woocommerce-admin/pull/4721#pullrequestreview-445959666) to address and then I think this one is ready to merge.

**Comment by octaedro** [comment]
Hi @psealock and @mattsherman, thank you for your review 🙂
Nice catch! You're right, we need to handle errors for `updateProfileItems`.
Taking a look at the code I realized we are not doing it in a couple of places ([here](https://github.com/woocommerce/woocommerce-admin/blob/main/client/profile-wizard/steps/theme/index.js#L99) and a few times in [this file](https://github.com/woocommerce/woocommerce-admin/blob/main/client/profile-wizard/index.js)).
If it's ok with you I will merge this PR and I'm going to create a follow-up PR to handle errors every time we use this method.

File: client/layout/controller.js
Comment (joshuatf) [inline client/layout/controller.js:185]: I think this should just use the `onboarding` feature flag.

```suggestion
	if ( window.wcAdminFeatures.onboarding ) {
```

File: config/plugin.json
Comment (joshuatf) [inline config/plugin.json:17]: I don't think we need to introduce a new feature flag, do we?

File: client/profile-wizard/index.js
Comment (joshuatf) [reply client/profile-wizard/index.js:226]: I think error handling might be good to handle before merging since this could end in being redirected back here if updating fails ( failure -> navigate to dashboard -> redirect back to profiler since not skipped ).

Maybe at least we could move the below `recordEvent` and `geHistory` calls inside `updateProfileItems().then( // ... here );` so they only fire on success.

**Review by joshuatf — COMMENTED** [review]
Fantastic work on this, @octaedro 🎉 

Really nice to have this update in place.  Left a couple (hopefully) quick comments that might be worth addressing before merge.

**Comment by octaedro** [comment]
Thank you @joshuatf for your review. I added the error handling in the commit aceb258.
Could you give this PR another look?

File: config/plugin.json
Comment (octaedro) [reply config/plugin.json]: Good idea! I removed it in the commits 9940b96 and 31c4154.

File: client/profile-wizard/index.js
Comment (mattsherman) [inline client/profile-wizard/index.js:236]: Are there any details in `response` when it fails that we could log? I'm trying to think about how an HE or dev could troubleshoot things if this situation was encountered. What are the most likely reasons why this would fail?

File: client/profile-wizard/index.js
Comment (mattsherman) [inline client/profile-wizard/index.js:243]:
```suggestion
						'There was a problem skipping the setup wizard.',
```

**Review by mattsherman — CHANGES_REQUESTED** [review]
Thanks for adding error handling. Left a few comment about the error handling.

File: src/API/OnboardingProfile.php
Comment (joshuatf) [inline src/API/OnboardingProfile.php:219]: Is the "skip" option only available on the "Store Details" step?  If so, I don't think we need to save the "step" data.

File: client/dashboard/index.js
Comment (joshuatf) [inline client/dashboard/index.js:48]: I think this should actually be:

```suggestion
			getHistory().push( getNewPath( {}, '/profiler', {} ) );
```

Otherwise the URL becomes `https://wp.local/wp-admin/https://wp.local/wp-admin/admin.php?page=wc-admin&path=/profiler&step=store-details` in my local install.

File: client/profile-wizard/index.js
Comment (joshuatf) [inline client/profile-wizard/index.js:234]: This one also needs to be updated:

```suggestion
					getHistory().push( getNewPath( {}, '/', {} ) );
```

File: client/homescreen/index.js
Comment (joshuatf) [inline client/homescreen/index.js:36]: This one should be updated as well.

**Review by joshuatf — COMMENTED** [review]
Left a few more clean-up comments I think need to be addressed before we can merge.

Also, want to note here that this breaks the profiler completion since we need to redirect to the dashboard after the profiler finishes, but as long as we get https://github.com/woocommerce/woocommerce-admin/pull/4717 merged in, this will be fixed.

File: client/profile-wizard/index.js
Comment (octaedro) [reply client/profile-wizard/index.js]: Hi, thank you for your suggestions. I removed the `throw` in the commit bec1b23

File: client/homescreen/index.js
Comment (octaedro) [reply client/homescreen/index.js]: Thank you @joshuatf, I did the updates in the commit 0a560f2

File: src/API/OnboardingProfile.php
Comment (octaedro) [reply src/API/OnboardingProfile.php:219]: I'm not sure if I understood correctly, I'm using the `step` to take the user back to the OBW step where they left off. If you think that this is not necessary just let me know and I'll change it 🙂

Comment (joshuatf) [reply src/API/OnboardingProfile.php:219]: I think we won't need it since the "Skip" button is only added to the "Store Details" step in https://github.com/woocommerce/woocommerce-admin/pull/4771/files

We can probably safely remove any reference to this saved step and "`lastStep` since it will always be the first step if they skip.

File: client/profile-wizard/index.js
Comment (joshuatf) [inline client/profile-wizard/index.js:245]: We won't need to check `response.status` here since this is already checked in the data actions.  The `.then()` block will only be triggered if the response was successful:

```suggestion
				recordEvent( 'storeprofiler_store_details_skip' );
				getHistory().push( getNewPath( {}, '/', {} ) );
```

File: client/profile-wizard/index.js
Comment (joshuatf) [inline client/profile-wizard/index.js:94]: I think this call can be removed.  Once they update `skipped` to `true` I don't think we need to revert it back to `false`.

It can remain `true` and they either complete the profiler or not.  Leaving `skipped` as is provides us more information.

**Review by joshuatf — COMMENTED** [review]
Clarified one of my previous suggestions and left a couple minor clean-up comments.  I think the test branch `add/4568_test` may also need a rebase.

File: client/profile-wizard/index.js
Comment (octaedro) [reply client/profile-wizard/index.js]: Sounds good to me. I removed it in the commit 4b73e58

File: client/profile-wizard/index.js
Comment (octaedro) [reply client/profile-wizard/index.js:237]: > We can probably safely remove any reference to this saved step and "lastStep since it will always be the first step if they skip.

Now I see, thank you @joshuatf, I removed it in the commit 37cab6

> We won't need to check response.status here since this is already checked in the data actions. The .then() block will only be triggered if the response was successful:

It seems like there is something tricky going on here, because with some types of errors the response will be `undefined` but in a "success" situation. So the user would be redirected to the home screen without receiving an alert.
The `catch` is triggered because the `response` doesn't have a `status`.
I created the branch `add/4568_test_2` with [the line commented](https://github.com/woocommerce/woocommerce-admin/pull/4721/files#diff-713853358fa08698c1bf0f8dfa42ca74R234) and adding a `die()` in the [update_items method](https://github.com/woocommerce/woocommerce-admin/blob/add/4568_test_2/src/API/OnboardingProfile.php#L129) if you want to test it.

Comment (joshuatf) [reply client/profile-wizard/index.js:237]: Ah, I see.  Good catch.  So after this line:
https://github.com/woocommerce/woocommerce-admin/blob/bd0d7dce14ebc3809f1d1bacd0155c07dc6345db/packages/data/src/onboarding/actions.js#L56

We're also missing: `throw new Error();` which will allow us to catch errors after `updateProfileItems()`.  Do you mind adding that here?  It seems to get things working for me with your demo branch, but  we should definitely check and make sure it doesn't break other areas of the onboarding.

Screen Shot 2020-07-15 at 6 39 42 PM

File: src/API/OnboardingProfile.php
Comment (joshuatf) [reply src/API/OnboardingProfile.php:219]: Just a heads up (not sure if you're still working on this) but this can be removed still.

File: client/profile-wizard/index.js
Comment (octaedro) [reply client/profile-wizard/index.js]: Got it! I added the code in the commit 3705935
I did a rebase and updated the test branch (`add/4568_test`). Now I'm testing and it looks like everything is working correctly now.
Could you take another look at this PR?

File: src/API/OnboardingProfile.php
Comment (joshuatf) [reply src/API/OnboardingProfile.php:219]: This property still needs to be removed.

File: tests/api/onboarding-profile.php
Comment (joshuatf) [inline tests/api/onboarding-profile.php:102]: These tests will need to be updated after `skip` is removed.

**Review by joshuatf — APPROVED** [review]
Thanks for all the effort on this, Fernando!  I know there was a lot of discussion and revision on this one, but this is a great improvement to the profiler.

There are a couple more issues around error handling and revisiting the profiler that we need to address, but I'll open some issues so we can tackle in a follow up.

Pre-approving pending the removal of the saved `step` property and test updates and then LGTM! 💯

File: tests/api/onboarding-profile.php
Comment (octaedro) [reply tests/api/onboarding-profile.php]: Nice catch @joshuatf! I removed the property and corrected the tests in the commit 1511792

**Pull request merged**
